qiskit_algorithms/custom_types.py

```diff
@@ -12,11 +12,11 @@
 
 """Types used by the qiskit-algorithms package."""
 
-from typing import Any, Protocol, Union
+from typing import Any, List, Protocol, Union
 
 from qiskit import QuantumCircuit
 
-_Circuits = Union[list[QuantumCircuit], QuantumCircuit]
+_Circuits = Union[List[QuantumCircuit], QuantumCircuit]
```

Comment:
You can use `list` in 3.8 - which arguably is better and what we have tried to do elsewhere. You do need to add a future import to have that work `from __future__ import annotations`. You can see examples in this repo e.g. in this file which does that and uses it https://github.com/qiskit-community/qiskit-algorithms/blob/main/qiskit_algorithms/time_evolvers/pvqd/pvqd_result.py

Reply:
I initially went for 
```python
_Circuits = list[QuantumCircuit] | QuantumCircuit
```
but then `make lint` complained with:
```
************* Module qiskit_algorithms.custom_types
qiskit_algorithms/custom_types.py:21:12: E1131: unsupported operand type(s) for | (unsupported-binary-operation)
```
which is why I went for `Union` instead of `|`. Or is there a workaround I don't know of?

Reply:
That future import is needed for `|` as well - but if you had that import then as I mentioned I was not not sure here as I recall having issues when defining types before using these newer aspects where we had to have them defined with the former Typing constructs. If you search in this repo you will find just a few Union hits where they are pretty much limited to being used in type defines like this - I also see List being used in the ones here so it may well be the same issue with that. I commented mostly as a saw the commit that changed things to fix it for lint in 3.8 from list to List that was all the change I did not see a from future import being removed which is needed to type things this newer way in 3.8.

Reply:
One further comment around 3.8. Qiskit has already deprecated support for Python 3.8 and support will be removed in 1.3.0 which is due in Nov. Python 3.8 EOL is Oct this year so among the changes here it could be a choice to drop 3.8 support too along with these changes. To drop this would evidently also include dropping it from CI tests etc and bumping things to run at 3.9 min. 

In talking about versions I will note 3.13 is planned to be available beginning of Oct. There is an issue on Qiskit Qiskit/qiskit#12903 to support this so again depending on timeline.... but hopefully that's just adding it as supported and to CI to test when dependencies are there and it just works!